File: app/src/main/java/org/schabi/newpipe/notifications/NotificationIcon.java
package org.schabi.newpipe.notifications;

import android.app.ActivityManager;
import android.content.Context;
import android.graphics.Bitmap;
import android.view.View;

import com.nostra13.universalimageloader.core.ImageLoader;
import com.nostra13.universalimageloader.core.assist.FailReason;
import com.nostra13.universalimageloader.core.assist.ImageSize;
import com.nostra13.universalimageloader.core.listener.SimpleImageLoadingListener;

import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.core.SingleEmitter;
import io.reactivex.rxjava3.core.SingleOnSubscribe;

final class NotificationIcon implements SingleOnSubscribe<Bitmap> {

    private final String url;
    private final int size;

    NotificationIcon(final Context context, final String url) {
        this.url = url;
        this.size = getIconSize(context);
    }

    @Override
    public void subscribe(@NonNull final SingleEmitter<Bitmap> emitter) throws Throwable {
        ImageLoader.getInstance().loadImage(
                url,
                new ImageSize(size, size),
                new SimpleImageLoadingListener() {

                    @Override
                    public void onLoadingFailed(final String imageUri,
                                                final View view,
                                                final FailReason failReason) {
                        emitter.onError(failReason.getCause());
                    }

                    @Override
                    public void onLoadingComplete(final String imageUri,
                                                  final View view,
                                                  final Bitmap loadedImage) {
                        emitter.onSuccess(loadedImage);
                    }
                }
        );
    }

    private static int getIconSize(final Context context) {
        final ActivityManager activityManager = (ActivityManager) context.getSystemService(
                Context.ACTIVITY_SERVICE
        );
        final int size2 = activityManager != null ? activityManager.getLauncherLargeIconSize() : 0;
        final int size1 = context.getResources()
                .getDimensionPixelSize(android.R.dimen.app_icon_size);
        return Math.max(size2, size1);
    }
}

File: app/src/main/java/org/schabi/newpipe/local/feed/notifications/NotificationIcon.kt
package org.schabi.newpipe.local.feed.notifications

import android.app.ActivityManager
import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import com.squareup.picasso.Picasso
import com.squareup.picasso.Target
import io.reactivex.rxjava3.core.SingleEmitter
import io.reactivex.rxjava3.core.SingleOnSubscribe
import org.schabi.newpipe.util.PicassoHelper

internal class NotificationIcon(
    context: Context,
    private val url: String,
) : SingleOnSubscribe<Bitmap> {

    private val size = getIconSize(context)

    override fun subscribe(emitter: SingleEmitter<Bitmap>) {
        val target = SingleEmitterTarget(emitter)
        PicassoHelper.loadThumbnail(url)
            .resize(size, size)
            .centerCrop()
            .into(target)
        emitter.setCancellable {
            PicassoHelper.cancelRequest(target)
        }
    }

    private class SingleEmitterTarget(private val emitter: SingleEmitter<Bitmap>) : Target {
        override fun onBitmapLoaded(bitmap: Bitmap, from: Picasso.LoadedFrom?) {
            if (!emitter.isDisposed) {
                emitter.onSuccess(bitmap)
            }
        }

        override fun onBitmapFailed(e: Exception, errorDrawable: Drawable?) {
            emitter.tryOnError(e)
        }

        override fun onPrepareLoad(placeHolderDrawable: Drawable?) = Unit
    }

    private companion object {

        fun getIconSize(context: Context): Int {
            val activityManager = context.getSystemService(
                Context.ACTIVITY_SERVICE
            ) as ActivityManager?
            val size1 = activityManager?.launcherLargeIconSize ?: 0
            val size2 = context.resources.getDimensionPixelSize(android.R.dimen.app_icon_size)
            return maxOf(size2, size1)
        }
    }
}
